pypred/ast.py

```diff
@@ -32,7 +32,11 @@ def __exit__(self, exc_type=None, exc_val=None, exc_tb=None):
         self.analyze = self._analyze
 
     def resolve_identifier(self, ident):
-        return self.pred.resolve_identifier(self.doc, ident)
+        res = self.pred.resolve_identifier(self.doc, ident)
+        # We always evaluate collections as set
+        if isinstance(res, list):
```

Comment:
Would this be required if both operands are already sets? In other words, is the _set_ equality operator already working?

Reply:
I'm just concerned that transforming lists to sets may be unexpected for users. If we make this choice, it will have to be clearly documented.

The alternatives could be : 
- to support only comparing sets, and making sure in our code that we always pass sets to pypred
- To also support declaring lists as litterals, thereby allowing comparisons of lists or sets.

Reply:
> Would this be required if both operands are already sets? In other words, is the set equality operator already working, just not documented?

If the identifier resolves as a set, the equality will work without this code, yes. 

> I'm just concerned that transforming lists to sets may be unexpected for users. If we make this choice, it will have to be clearly documented.

You can only define set Literals atm, no list Literals, so you just can't compare list (except for contains) with the library now. In the README, :
> It supports the following:
> Set literal, containing string literals, numeric and constant values

I agree that we should update the documentation to reflect that if you compare against a list, it will be evaluated as a set, but I don't feel like it will be confusing since it's not available now.


> to support only comparing sets, and making sure in our code that we always pass sets to pypred

Isn't this the same as casting the list to a set in the lib, but only for us vs benefiting the whole community?

> To also support declaring lists as literals, thereby allowing comparisons of lists or sets.

That's an option, but that would be more work for not much in my opinion. I think that the idea for supporting only sets is related to performance (contains is way faster because set is implemented as a hash table)

Reply:
> Isn't this the same as casting the list to a set in the lib, but only for us vs benefiting the whole community?

Yes, but introducing this in the language is hard to reverse. It blocks a potential future support for actual lists. What if there is a use case where order and duplicates is important? If we introduce the notion that [1,2,2,3] == [2,1,3], we can't change it without introducing a breaking change.

For us, we know that we only care about comparing sets, so in fact, it is our error to store elements in lists, which should be in sets. If we fix this error in BM, it will just work with pypred.

If at some point we do need to compare lists (i.e need order/duplicates) then we can add support for list litterals in pypred.

Reply:
Alright, that's a good argument. Will cancel the PR and move the logic to BM :)

Reply:
Sorry about pooping on PR #1 🙈

Reply:
It was a rehearsal for the others to come 😅  I thought I would start with this one since it was the "simplest" one. 🙈   

The other operators should be less "prone to discussion" since they don't have the potential to introduce breaking changes 😝 🤞